lib/simulator_avx.h

```diff
@@ -32,8 +32,9 @@ class SimulatorAVX final {
   using State = typename StateSpace::State;
   using fp_type = typename StateSpace::fp_type;
 
-  SimulatorAVX(unsigned num_qubits, unsigned num_threads)
-      : num_qubits_(num_qubits), num_threads_(num_threads) {}
+  template <typename... Args>
+  explicit SimulatorAVX(unsigned num_qubits, Args&&... args)
```

Comment:
Could we pass a pre-constructed For object to this constructor instead? That would give users the flexibility to configure their For object outside its constructor (e.g. if some parts of the For object could generate errors during setup).

If this makes tests overly verbose, a default value for the For argument could be used.

Reply:
Likewise for other simulators and statespace objects.

Reply:
One can pass a pre-constructed For object to this constructor. The following will work:
```C++
MyFor my_for(num_threads, ...);
SimulatorAVX<const MyFor&> simulator(num_qubits, my_for);
```
If `MyFor` has a copy constructor, the following will also work:
```C++
MyFor my_for(num_threads, ...);
SimulatorAVX<MyFor> simulator(num_qubits, my_for);